pgdist.1
.TH man 1 "31 October 2010" "1.0" "PGdist manual"

.SH NAME
PGdist - create, develop and manage your PostgreSQL projects

.SH SYNOPSIS
.TP
.B pgdist
[-v|--verbose]
[-?|--help]
[--git-diff]
[--less]
[--noless]
[--all]
[-f|--force]
[-c|--config <\fIfile\fR>]
[--color <[\fIauto\fR, \fInever\fR, \fIalways\fR]>
[--swap]
[--gitversion]
[--no-owner]
[--no-acl]
[--diff-raw]
[-w|--ignore-all-space]
[--no-clean]
[--cache]
[--pg_extractor]
[--pg_extractor_basedir <\fIdirectory\fR>]
[--pre-load <\fIfile\fR>]
[--post-load <\fIfile\fR>]
[--pre-remoted-load <\fIfile\fR>]
[--post-remoted-load <\fIfile\fR>]
[--pre-load-old <\fIfile\fR>]
[--post-load-old <\fIfile\fR>]
[--pre-load-new <\fIfile\fR>]
[--post-load-new <\fIfile\fR>]
[--showall]
[-d|--dbname <\fIdbname\fR>]
[-h|--host <\fIhost\fR>]
[-p|--port <\fIport\fR>]
[-U|--username <\fIusername\fR>]
[-C|--create]
[--directory <\fIpath\fR>]
[--syslog-facility <\fIfacility\fR>]
[--syslog-ident <\fIident\fR>]
<\fIcommand\fR> [<\fIargs\fR>]

.SH DESCRIPTION
PGdist is program for PostgreSQL projects management from development to production.

PGdist commands are divided to two sections, develop and management.
.TP
PGdist can:
- manage source files, collaborating with Git

- test project installation in postgres

- help to create update scripts and test it for mistakes

- display project‘s diff and existing database structure

- watch for proper role creation in database

- allow dependency settings

- compare data in tables

- make installations and updates on production server (inc. roles, dependencies)

- allow to split install and update scripts to parts, intended to be or not to be done in particular transaction

- show diffs in structured and colored mode
.TP
PGdist requires:
- Postgres

- Git

.SH OPTIONS
.TP
\fB-v\fR, \fB--verbose\fR
Increase output verbosity (more detailed print).
.TP
\fB-?\fR, \fB--help\fR
Shows PGdist help with all commands and options.
.TP
\fB--git-diff\fR
Generates update from old projects from diff between files not pg_dump. This means that actual files are compared, not data from pg_dump.
.TP
\fB--less\fR
Prints output in less.

Warning: if used and PGdist will fail, you will loose error output.
.TP
\fB--noless\fR
Will not print output in less.
.TP
\fB--all\fR
Use all files. Used with command \fBadd\fR and \fBrm\fR.
.TP
\fB-f\fR, \fB--force\fR
Overwrite and remove files.
.TP
\fB-c\fR, \fB--config\fR \fI<file>\fR
Path to configuration file which will be used instead of standard configuration file.
.TP
\fB--color\fR <[\fIauto\fR, \fInever\fR, \fIalways\fR]>
Set when should be used colorred output (useful for diff commands).
.TP
\fB--swap\fR
Swap compare data.
.TP
\fB--gitversion\fR <\fIversion\fR>
Used as version for git tag.
.TP
\fB--no-owner\fR
Do not dump and compare owners.
.TP
\fB--no-acl\fR
Do not dump and compare access privileges (grant/revoke commands).
.TP
\fB--diff-raw\fR
Compare raw SQL dumps -> PGdist will not try to parse dumps.
.TP
\fB-w\fR, \fB--ignore-all-space\fR
When comparing, ignore all space diffrences.
.TP
\fB--no-clean\fR
Will not clean test database after load or update test.
.TP
\fB--cache\fR
Will cache remote dump for 4 hours.
.TP
\fB--pg_extractor\fR
PGdist will dump by pg_extractor, compare by diff -r.
.TP
\fB--pg_extractor_basedir\fR <\fIdirectory\fR>
PGdist will dump by pg_extractor do directory PG_EXTRACTOR_BASE_DIR.
.TP
\fB--pre-load\fR <\fIfile\fR>
SQL file to load before project.

Useful for cases where you are missing for example schema in your project, but have some tables in mentioned schema, so you will just preload file with that schema.
.TP
\fB--post-load\fR <\fIfile\fR>
SQL file to load after project.
.TP
\fB--pre-remoted-load\fR <\fIfile\fR>
SQL file to load before old version of project.
.TP
\fB--post-remoted-load\fR <\fIfile\fR>
SQL file to load after old version of project.
.TP
\fB--pre-load-new\fR <\fIfile\fR>
SQL file to load before new version of project.
.TP
\fB--post-load-new\fR <\fIfile\fR>
SQL file to load after new version of project.
.TP
\fB--showall\fR
Show all versions. Used with \fBlist\fR command.
.TP
\fB-d\fR, \fB--dbname\fR <\fIdbname\fR>
Specifies the name fo the database to connect to.
.TP
\fB-h\fR, \fB--host\fR <\fIhost\fR>
Specifies the host name of the machine on which the server is running.
.TP
\fB-p\fR, \fB--post\fR <\fIport\fR>
Specifies the TCP port or the local Unix-domain socket file.
.TP
\fB-U\fR, \fB--username\fR <\fIusername\fR>
Connect to the database with username.
.TP
\fB-C\fR, \fB--create\fR
Create database if it does not exist. Used with \fBinstall\fR command.
.TP
\fB--directory\fR <\fIdirectory\fR>
Directory containing scripts for install and update.
.TP
\fB--syslog-facility\fR <\fIfacility\fR>
Syslog facility.
.TP
\fB--syslog-ident\fR <\fIident\fR>
Syslog ident.

.SH DEVELOP COMMANDS
.TP
\fBpgdist init\fR <\fIPROJECT\fR> [\fIPATH\fR]
initialize pgdist \fIPROJECT\fR in \fIPATH\fR
.TP
\fBpgdist create-schema\fR <\fISCHEMA\fR>
create new \fISCHEMA\fR directory structure
.TP
\fBpgdist status\fR
show new files and removed files compared to pg_project.sql
.TP
\fBpgdist add\fR [\fIFILE\fR ...]
add files to pg_project.sql
.TP
\fBpgdist rm\fR [\fIFILE\fR ...]
remove deleted files from pg_project.sql
.TP
\fBpgdist part-add\fR [\fBnot-single-transaction\fR]
add new part with single or not single (if specified) transaction to pg_project.sql
.TP
\fBpgdist part-rm\fR <\fIPART_NUMBER\fR>
remove part from pg_project.sql, part files are added to previous part, use -f to remove files from pg_project.sql too
.TP
\fBpgdist test-load\fR
load project to testing postgres
.TP
\fBpgdist create-version\fR <\fIVERSION\fR> [\fIGIT_TAG\fR]
create version files
.TP
\fBpgdist create-update\fR <\fIGIT_TAG NEW_VERSION\fR> [\fIPARTS\fR]
create update files with differencies

\fIGIT_TAG\fR - old version tag

\fINEW_VERSION\fR - new version

\fIPARTS\fR - number of parts you want to create
.TP
\fBpgdist part-update-add\fR <\fIOLD_VERSION\fR> <\fINEW_VERSION\fR> [\fBnot-single-transaction\fR]
add update part file
.TP
\fBpgdist part-update-rm\fR <\fIOLD_VERSION\fR> <\fINEW_VERSION\fR> <\fIPART_NUMBER\fR>
delete update part file
.TP
\fBpgdist test-update\fR <\fIGIT_TAG NEW_VERSION\fR>
load old and new version and compare it

\fIGIT_TAG\fR - old version tag

\fINEW_VERSION\fR - new version
.TP
\fBpgdist diff-db\fR <\fIPGCONN\fR> [\fIGIT_TAG\fR]
diff project and database
.TP
\fBpgdist diff-db-file\fR <\fIPGCONN\fR> <\fIFILE\fR>
diff file and database
.TP
\fBpgdist diff-file-db\fR <\fIFILE\fR> <\fIPGCONN\fR>
diff database and file
.TP
\fBpgdist role-list\fR
print roles in project
.TP
\fBpgdist role-add\fR <\fINAME\fR> [\fBlogin\fR|\fBnologin\fR] [\fBpassword\fR]
add role to project
.TP
\fBpgdist role-change\fR <\fINAME\fR> [\fBlogin\fR|\fBnologin\fR] [\fBpassword\fR]
change role
.TP
\fBpgdist role-rm\fR <\fINAME\fR>
remove role from project, not remove from databases
.TP
\fBpgdist require-add\fR <\fIPROJECT\fR> <\fIGIT\fR> <\fIGIT_TREE_ISH\fR>
add require to another project

\fI<GIT_TREE_ISH>\fR - indicates a tree, commit or tag object name of required project
.TP
\fBpgdist require-rm\fR <\fIPROJECT\fR>
remove require to another project
.TP
\fBpgdist dbparam-set\fR [\fIPARAM\fR ...]
parameters with create a database (e.g.: OWNER lbadmin ...)
.TP
\fBpgdist dbparam-get\fR
print parameters to create a database
.TP
\fBpgdist data-add\fR <\fITABLE\fR> [\fICOLUMN1\fR ...]
add table to compare data
.TP
\fBpgdist data-rm\fR <\fITABLE\fR>
remove table to compare data
.TP
\fBpgdist data-list\fR
list table of data compare
.RE

.SH MANAGEMENT COMMANDS
.TP
\fBpgdist list\fR [\fIPROJECT\fR [\fIDBNAME\fR]]
show list of installed projects in databases
.TP
\fBpgdist install\fR <\fIPROJECT\fR> <\fIDBNAME\fR> [\fIVERSION\fR]
install project to database
.TP
\fBpgdist check-update [\fIPROJECT\fR [\fIDBNAME\fR [\fIVERSION\fR]]]
check update project
.TP
\fBpgdist update [\fIPROJECT\fR [\fIDBNAME\fR [\fIVERSION\fR]]]
update project
.TP
\fBpgdist clean\fR <\fIPROJECT\fR> [\fIDBNAME\fR]
remove all info about project
.TP
\fBpgdist set-version\fR <\fIPROJECT\fR> <\fIDBNAME\fR> <\fIVERSION\fR>
force change version without run scripts
.TP
\fBpgdist get-version\fR <\fIPROJECT\fR> <\fIDBNAME\fR>
print installed version of project
.TP
\fBpgdist pgdist-update [\fIDBNAME\fR]
update pgdist version in database
.TP
\fBpgdist log [\fIPROJECT\fR [\fIDBNAME\fR]]
print history of installed projects

.SH DEVELOP CONFIG FILE
Configuration file is located at `~/.pgdist`.
.TP
It looks like this:
[pgdist]
test_db: pgdist@sqltest/postgres

\fBtest_db\fR - PG connection to testing postgres database.

.SS PGCONN
It defines ssh connection (\fBnot required\fR) + connection URI.

Please use connection URI \fBwithout\fR \fIpostgresql://\fR string.

If you choose to use ssh connection, it is highly recommended to set up \fBssh-key\fR.

See more about connection URI: \fIhttps://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING\fR.
.TP
Examples:
localhost/test_database

postgres@/test_database

localhost//postgres@/
.TP
Below PGCONN will define ssh connection to \fBmy_server\fR with user \fBroot\fR, then open postgres connection with PG user \fBpostgres\fR, password \fBPASSWORD\fR, PG host \fBlocalhost\fR, listening port \fB5042\fR, database \fBtest_database\fR and connection timeout \fB10 seconds\fR.

root@my_server//postgres:PASSWORD@localhost:5042/test_database?connection_timetout=10

.SH DISTRIBUTION CONFIG FILE
Configuration file is located at `/etc/pgdist.conf`.
.TP
It looks like this:
[pgdist]

installation_path = /usr/share/pgdist/install

password_path = /etc/lbox/postgresql/roles

pguser = postgres

pgdatabase = postgres

pghost = localhost

pgport = 5432

.TP
\fBinstallation_path\fR - path to version/updates scripts
.TP
\fBpassword_path\fR - path to roles passwords
.TP
\fBpguser\fR - default PG user to connect with
.TP
\fBpgdatabase\fR - optional - name of database to connect to
.TP
\fBpghost\fR - optional - PG host
.TP
\fBpgport\fR - optional - port that PG listens to

.SH SEE ALSO
\fIhttps://github.com/linuxbox-cz/pgdist\fR

.SH BUGS
No known bugs.

.SH AUTHOR
Marian Krucina \fIhttps://www.LinuxBox.cz/\fR

Tadeáš Popov \fIhttps://github.com/TadeasPopov\fR


src/pgdist.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

import os
import sys
import io
import argparse
import subprocess
import logging
import logging.handlers

reload(sys)
sys.setdefaultencoding('utf-8')

description = """
PGdist - distributes PotgreSQL functions, tables, etc...
PGdist Devel - develop PostgreSQL project

    init PROJECT [PATH] - initialize pgdist PROJECT in PATH
    create-schema SCHEMA - create new SCHEMA directory structure
    status - show new files and removed files compared to pg_project.sql
    add FILE1 [FILE2 ...] - add files to pg_project.sql
    rm FILE1 [FILE2 ...] - removed files from pg_project.sql
    part-add [not-single-transaction] - add new part with single or not single (if specified) transaction to pg_project.sql
    part-rm PART_NUMBER - remove part from pg_project.sql, part files are added to previous part, use -f to remove files from pg_project.sql too

    test-load - load project to testing postgres
    create-version VERSION [GIT_TAG] - create version files
    create-update GIT_TAG NEW_VERSION [PARTS] - create update files with differencies
                                          - GIT_TAG - old version tag
                                          - NEW_VERSION - new version
                                          - PARTS - number of parts you want to create
    part-update-add OLD_VERSION NEW_VERSION [not-single-transaction] - add update part file
    part-update-rm OLD_VERSION NEW_VERSION PART_NUMBER - delete update part file
    test-update GIT_TAG NEW_VERSION - load old and new version and compare it
                                          - GIT_TAG - old version tag
                                          - NEW_VERSION - new version

    diff-db PGCONN [GIT_TAG] - diff project and database
    diff-db-file PGCONN FILE - diff file and database
    diff-file-db FILE PGCONN - diff database and file

    role-list - print roles in project
    role-add NAME [login|nologin] [password] - add role to project
    role-change NAME [login|nologin] [password] - change role
    role-rm NAME - remove role from project, not remove from databases

    require-add PROJECT GIT GIT_TREE_ISH - add require to another project
                                          - GIT_TREE_ISH - indicates a tree, commit or tag object name of required project
    require-rm PROJECT - remove require to another project

    dbparam-set [PARAM [...]] - parameters with create a database (e.g.: OWNER lbadmin ...)
    dbparam-get - print parameters to create a database

    data-add TABLE [COLUMN1 [...]] - add table to compare data
    data-rm TABLE - remove table to compare data
    data-list - list table of data compare

PGdist Server - manage projects in PostgreSQL database

    list [PROJECT [DBNAME]] - show list of installed projects in databases
    install PROJECT DBNAME [VERSION] - install project to database
    check-update [PROJECT [DBNAME [VERSION]]] - check update project
    update [PROJECT [DBNAME [VERSION]]] - update project
    clean PROJECT [DBNAME] - remove all info about project
    set-version PROJECT DBNAME VERSION - force change version without run scripts
    get-version PROJECT DBNAME - print installed version of project
    pgdist-update [DBNAME] - update pgdist version in database
    log [PROJECT [DBNAME]] - print history of installed projects

PGCONN - ssh connection + connection URI, see:
    https://www.postgresql.org/docs/current/static/libpq-connect.html#LIBPQ-CONNSTRING
    without string 'postgresql://'
    examples:
        localhost/mydb - connect to mydb
        root@server//user@/mydb - connect to server via ssh and next connect to postgres as user into mydb

Develop configuration:
    connection to testing database in file "~/.pgdist" (or ".pgdist" in project path) with content:
        [pgdist]
        test_db: user@host/dbname

        test_db - PGCONN to testing postgres, user has to create databases and users
Distribution configuration:
    Configuration file is located at `/etc/pgdist.conf`.
        [pgdist]
        installation_path = /usr/share/pgdist/install
        password_path = /etc/lbox/postgresql/roles
        pguser = postgres
        pgdatabase = pg_database
        pghost = localhost
        pgport = 5042

        pgdatabase, pghost and pgport are optional (does not have to be setted)
"""

def main():
	logging.basicConfig(format="%(message)s")
	parser = argparse.ArgumentParser(formatter_class=argparse.RawTextHelpFormatter, description=description, add_help=False)
	less = False

	# common argument
	parser.add_argument("-v", "--verbose", help="increase output verbosity", action="count", default=0)
	parser.add_argument("-?", "--help", dest="help", action="store_true", help="show this help message and exit")
	parser.add_argument("cmd", nargs="?")
	parser.add_argument("args", nargs="*")

	# develop projects
	parser.add_argument("--git-diff", dest="git_diff", help="generate diff against git files", action="store_true", default=False)
	parser.add_argument("--less", help="print output in less", action="store_true")
	parser.add_argument("--noless", help="don't print output in less", action="store_true")
	parser.add_argument("--all", help="use all files", action="store_true")
	parser.add_argument("-f", "--force", help="overwriting and removing files", action="store_true")
	parser.add_argument("-c", "--config", help="configuration file")
	parser.add_argument("--color", help="never, always or auto colorred output", default="auto", choices=["auto", "never", "always"])
	parser.add_argument("--swap", help="Swap compare data, commands diff-...", default=False, action="store_true")
	parser.add_argument("--gitversion", help="use as version for git tag")
	parser.add_argument("--no-owner", dest="no_owner", help="do not dump and compare ownership of objects", action="store_true")
	parser.add_argument("--no-acl", dest="no_acl", help="do not dump and compare access privileges (grant/revoke commands)", action="store_true")
	parser.add_argument("--diff-raw", dest="diff_raw", help="compare raw SQL dumps", action="store_true")
	parser.add_argument("-w", "--ignore-all-space", dest="ignore_space", help="ignore all white space", action="store_true", default=False)
	parser.add_argument("--no-clean", dest="no_clean", help="no clean test database after load/update test", action="store_true", default=False)
	parser.add_argument("--cache", dest="cache", help="cache dump remote database for 4 hours", action="store_true", default=False)
	parser.add_argument("--pg_extractor", dest="pg_extractor", help="Dump by pg_extractor, compare by diff -r", action="store_true")
	parser.add_argument("--pg_extractor_basedir", dest="pg_extractor_basedir", help="Dump by pg_extractor do directory PG_EXTRACTOR_BASEDIR")
	parser.add_argument("--pre-load", dest="pre_load", help="SQL file to load before load project")
	parser.add_argument("--post-load", dest="post_load", help="SQL file to load after load project")
	parser.add_argument("--pre-remoted-load", dest="pre_remoted_load", help="SQL file to load before load remote dump, command: diff-db")
	parser.add_argument("--post-remoted-load", dest="post_remoted_load", help="SQL file to load after project, command: diff-db, path unversion install")
	parser.add_argument("--pre-load-old", dest="pre_load_old", help="SQL file to load before load old version of the project")
	parser.add_argument("--post-load-old", dest="post_load_old", help="SQL file to load after load old version of the project")
	parser.add_argument("--pre-load-new", dest="pre_load_new", help="SQL file to load before load new version of the project")
	parser.add_argument("--post-load-new", dest="post_load_new", help="SQL file to load after load new version of the project")

	# install projects
	parser.add_argument("--showall", help="show all versions", action="store_true")
	parser.add_argument("-d", "--dbname", dest="dbname", help="Specifies the name of the database to connect to.")
	parser.add_argument("-h", "--host", dest="host", help="Specifies the host name of the machine on which the server is running.")
	parser.add_argument("-p", "--port", dest="port", help="Specifies the TCP port or the local Unix-domain socket file.")
	parser.add_argument("-U", "--username", dest="user", help="Connect to the database as the user username.")
	parser.add_argument("-C", "--create", dest="create", help="Create the database.", action="store_true")
	parser.add_argument("--directory", help="directory contains script install and update")
	parser.add_argument("--syslog-facility", dest="syslog_facility", help="syslog facility")
	parser.add_argument("--syslog-ident", dest="syslog_ident", help="syslog ident")

	args = parser.parse_args()
	less = None

	if args.help:
		parser.print_help()
		sys.exit(1)

	VERBOSE = logging.DEBUG + 5
	logging.addLevelName(VERBOSE, "VERBOSE")
	def verbose(message, *args, **kws):
		if logging.getLogger().root.isEnabledFor(VERBOSE):
			logging.getLogger().root._log(VERBOSE, message, args, **kws) 
	logging.verbose = verbose
	if args.verbose > 2:
		logging.basicConfig(format="%(asctime)-15s %(filename)s:%(lineno)d %(message)s")
		logging.getLogger().setLevel(logging.DEBUG)
		logging.debug("verbosity debug turned on")
	elif args.verbose > 0:
		logging.getLogger().setLevel(VERBOSE)
		logging.verbose("verbosity turned on")
	else:
		logging.getLogger().setLevel(logging.INFO)

	if args.syslog_facility or args.syslog_ident:
		handler = logging.handlers.SysLogHandler(facility=args.syslog_facility, address='/dev/log')
		if args.syslog_ident:
			handler.setFormatter(logging.Formatter(args.syslog_ident+": %(message)s"))
		logging.getLogger().addHandler(handler)

	if args.cmd in ("init", "create-schema", "status", "test-load", "create-version", "add", "rm",
		"part-add", "part-rm", "create-update", "test-update",
		"part-update-add", "part-update-rm",
		"diff-db", "diff-db-file", "diff-file-db",
		"role-list", "role-add", "role-change", "role-rm",
		"require-add", "require-rm", "dbparam-set", "dbparam-get",
		"data-add", "data-rm", "data-list"):

		sys.path.insert(1, os.path.join(sys.path[0], "dev"))
		import color
		import address
		import config
		import pg_parser
		import pg_project
		import pg_extractor as pg_extractor_m

		if args.less:
			less = True
		elif args.noless or args.cmd not in ("diff-db", "diff-db-file", "diff-file-db"):
			less = False
		else:
			less = sys.stdout.isatty()

		if less:
			buffer = io.StringIO()
			stdout = sys.stdout
			sys.stdout = buffer

		config.load(args.config)
		if less and args.color == "auto":
			color.set("always")
		else:
			color.set(args.color)

		config.git_diff = args.git_diff

	if args.cmd in ("list", "install", "check-update", "update", "clean", "set-version", "get-version","pgdist-update", "log"):
		sys.path.insert(1, os.path.join(sys.path[0], "mng"))
		import config
		import conninfo
		import pg_project

		config.load(args.config)

		if not args.user:
			args.user = config.get_pguser()
		if not args.dbname:
			args.dbname = config.get_pgdatabase()
		if not args.host:
			args.host = config.get_pghost()
		if not args.port:
			args.port = config.get_pgport()

	if args.pg_extractor:
		pg_extractor = pg_extractor_m.PG_extractor(args.pg_extractor_basedir)
	else:
		pg_extractor = None

	# develop projects
	if args.cmd == "init" and len(args.args) in (1, 2,):
		(project_name, directory) = args_parse(args.args, 2)
		if not directory:
			directory = os.getcwd()
		pg_project.project_init(project_name, directory)

	elif args.cmd == "create-schema" and len(args.args) in (1,):
		(schema_name,) = args_parse(args.args, 1)
		pg_project.create_schema(schema_name)

	elif args.cmd == "status" and len(args.args) in (0,):
		pg_project.status()

	elif args.cmd == "add":
		pg_project.add(args.args, args.all)

	elif args.cmd ==  "rm":
		pg_project.rm(args.args, args.all)

	elif args.cmd == "part-add" and len(args.args) in (0,1):
		(transaction_type,) = args_parse(args.args, 1)
		pg_project.part_add(transaction_type)

	elif args.cmd == "part-rm" and len(args.args) in (1,):
		(number,) = args_parse(args.args, 1)
		pg_project.part_rm(int(number), args.force)

	elif args.cmd == "test-load" and len(args.args) in (0,):
		pg_project.test_load(not args.no_clean, args.pre_load, args.post_load, pg_extractor=pg_extractor, no_owner=args.no_owner)

	elif args.cmd == "create-version" and len(args.args) in (1, 2,):
		(version, git_tag) = args_parse(args.args, 2)
		pg_project.create_version(version, git_tag, args.force)

	elif args.cmd == "create-update" and len(args.args) in (2, 3):
		(git_tag, new_version, part_count) = args_parse(args.args, 3)
		pg_project.create_update(git_tag, new_version, args.force, args.gitversion, clean=not args.no_clean, pre_load=args.pre_load, post_load=args.post_load,
			pre_load_old=args.pre_load_old, pre_load_new=args.pre_load_new, post_load_old=args.post_load_old, post_load_new=args.post_load_new,
			part_count=part_count or 1)

	elif args.cmd == "part-update-add" and len(args.args) in (2, 3):
		(old_version, new_version, transaction_type) = args_parse(args.args, 3)
		pg_project.part_update_add(old_version, new_version, transaction_type)

	elif args.cmd == "part-update-rm" and len(args.args) in (3,):
		(old_version, new_version, number) = args_parse(args.args, 3)
		pg_project.part_update_rm(old_version, new_version, int(number))

	elif args.cmd == "test-update" and len(args.args) in (2,):
		(git_tag, new_version) = args_parse(args.args, 1)
		pg_project.test_update(git_tag, new_version, gitversion=args.gitversion, clean=not args.no_clean, pre_load=args.pre_load, post_load=args.post_load,
			pre_load_old=args.pre_load_old, pre_load_new=args.pre_load_new, post_load_old=args.post_load_old, post_load_new=args.post_load_new,
			pg_extractor=pg_extractor, no_owner=args.no_owner)

	elif args.cmd == "diff-db" and len(args.args) in (1, 2):
		(pgconn, git_tag) = args_parse(args.args, 2)
		pg_project.diff_pg(address.Address(pgconn), git_tag, args.diff_raw, not args.no_clean, args.no_owner, args.no_acl,
			pre_load=args.pre_load, post_load=args.post_load, pre_remoted_load=args.pre_remoted_load, post_remoted_load=args.post_remoted_load,
			swap=args.swap, pg_extractor=pg_extractor, cache=args.cache, ignore_space=args.ignore_space)

	elif args.cmd == "diff-db-file" and len(args.args) in (2,):
		(pgconn, file) = args_parse(args.args, 2)
		pg_project.diff_pg_file(address.Address(pgconn), file, args.diff_raw, not args.no_clean, args.no_owner, args.no_acl,
			pre_load=args.pre_load, post_load=args.post_load, pre_remoted_load=args.pre_remoted_load, post_remoted_load=args.post_remoted_load,
			swap=args.swap, pg_extractor=pg_extractor, cache=args.cache, ignore_space=args.ignore_space)

	elif args.cmd == "diff-file-db" and len(args.args) in (2,):
		(file, pgconn) = args_parse(args.args, 2)
		pg_project.diff_pg_file(address.Address(pgconn), file, args.diff_raw, not args.no_clean, args.no_owner, args.no_acl,
			pre_load=args.pre_load, post_load=args.post_load, pre_remoted_load=args.pre_remoted_load, post_remoted_load=args.post_remoted_load,
			swap=not args.swap, pg_extractor=pg_extractor, cache=args.cache, ignore_space=args.ignore_space)

	elif args.cmd == "role-list" and len(args.args) in (0,):
		pg_project.role_list()

	elif args.cmd == "role-add" and len(args.args) in (1, 2, 3,):
		(name, arg1, arg2) = args_parse(args.args, 3)
		pg_project.role_add(name, arg1, arg2)

	elif args.cmd == "role-change" and len(args.args) in (1, 2, 3,):
		(name, arg1, arg2) = args_parse(args.args, 3)
		pg_project.role_change(name, arg1, arg2)

	elif args.cmd == "role-rm" and len(args.args) in (1,):
		(name,) = args_parse(args.args, 1)
		pg_project.role_rm(name)

	elif args.cmd == "require-add" and len(args.args) in (3,):
		(project_name, git, git_tree_ish) = args_parse(args.args, 3)
		pg_project.require_add(project_name, git, git_tree_ish)

	elif args.cmd == "require-rm" and len(args.args) in (1,):
		(project_name, ) = args_parse(args.args, 1)
		pg_project.require_rm(project_name)

	elif args.cmd == "dbparam-set":
		if len(args.args) == 0:
			dbparam = None
		else:
			dbparam = " ".join(args.args)
		pg_project.dbparam_set(dbparam)

	elif args.cmd == "dbparam-get" and len(args.args) in (0,):
		pg_project.dbparam_get()

	elif args.cmd == "data-add" and len(args.args) >= 1:
		table = args.args[0]
		columns = args.args[1:]
		pg_project.tabledata_add(table, columns)

	elif args.cmd == "data-rm" and len(args.args) in (1,):
		(table, ) = args_parse(args.args, 1)
		pg_project.tabledata_rm(table)

	elif args.cmd == "data-list" and len(args.args) in (0,):
		pg_project.tabledata_list()

	# install projects
	elif args.cmd == "list" and len(args.args) in (0, 1, 2,):
		(project_name, dbname) = args_parse(args.args, 2)
		pg_project.prlist(project_name, dbname or args.dbname, conninfo.ConnInfo(args), args.directory or config.get_install_path(), args.showall)

	elif args.cmd == "install" and len(args.args) in (2, 3,):
		(project_name, dbname, version) = args_parse(args.args, 3)
		pg_project.install(project_name, dbname, version, conninfo.ConnInfo(args), args.directory or config.get_install_path(), args.create, False)

	elif args.cmd == "check-update" and len(args.args) in (0, 1, 2, 3,):
		(project_name, dbname, version) = args_parse(args.args, 3)
		pg_project.check_update(project_name, dbname or args.dbname, version, conninfo.ConnInfo(args), args.directory or config.get_install_path())

	elif args.cmd == "update" and len(args.args) in (0, 1, 2, 3,):
		(project_name, dbname, version) = args_parse(args.args, 3)
		pg_project.update(project_name, dbname or args.dbname, version, conninfo.ConnInfo(args), args.directory or config.get_install_path())

	elif args.cmd == "clean" and len(args.args) in (1, 2,):
		(project_name, dbname) = args_parse(args.args, 2)
		pg_project.clean(project_name, dbname or args.dbname, conninfo.ConnInfo(args))

	elif args.cmd == "set-version" and len(args.args) in (3,):
		(project_name, dbname, version) = args_parse(args.args, 3)
		pg_project.set_version(project_name, dbname, version, conninfo.ConnInfo(args))

	elif args.cmd == "get-version" and len(args.args) in (2,):
		(project_name, dbname) = args_parse(args.args, 2)
		pg_project.get_version(project_name, dbname, conninfo.ConnInfo(args))

	elif args.cmd == "pgdist-update" and len(args.args) in (0,1):
		(dbname,) = args_parse(args.args, 1)
		pg_project.pgdist_update(dbname, conninfo.ConnInfo(args))

	elif args.cmd == "log" and len(args.args) in (0, 1, 2,):
		(project_name, dbname) = args_parse(args.args, 2)
		pg_project.history(project_name, dbname or args.dbname, conninfo.ConnInfo(args))

	else:
		parser.print_help()
		sys.exit(1)

	if less:
		pager = subprocess.Popen(["less", "-FKSMIR"], bufsize=1, stdin=subprocess.PIPE, stdout=stdout)
		pager.stdin.write(buffer.getvalue())
		pager.stdin.close()
		pager.wait()

	sys.exit(0)

def args_parse(args, n):
	return args + [None] * (n - len(args))

if __name__ == "__main__":
	try:
		main()
	except KeyboardInterrupt, e:
		pass
